<a href="https://colab.research.google.com/github/Nageswar-Sahoo/S5/blob/main/EVA7_Assignment_S5_File_9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Libraries

In [ ]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms

## Data Transformations

We first start with defining our data transformations. We need to think what our data is and how can we augment it to correct represent images which it might not see otherwise. 


In [ ]:
# Train Phase transformations
train_transforms = transforms.Compose([
                                      #  transforms.Resize((28, 28)),
                                      #  transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1),
                                       transforms.RandomRotation((-7.0, 7.0), fill=(1,)),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,)) # The mean and std have to be sequences (e.g., tuples), therefore you should add a comma after the values. 
                                       # Note the difference between (0.1307) and (0.1307,)
                                       ])

# Test Phase transformations
test_transforms = transforms.Compose([
                                      #  transforms.Resize((28, 28)),
                                      #  transforms.ColorJitter(brightness=0.10, contrast=0.1, saturation=0.10, hue=0.1),
                                       transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])


# Dataset and Creating Train/Test Split

In [ ]:
train = datasets.MNIST('./data', train=True, download=True, transform=train_transforms)
test = datasets.MNIST('./data', train=False, download=True, transform=test_transforms)

  0%|          | 0/9912422 [00:00<?, ?it/s]

Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw



  0%|          | 0/28881 [00:00<?, ?it/s]

Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw



  0%|          | 0/1648877 [00:00<?, ?it/s]

Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw



  0%|          | 0/4542 [00:00<?, ?it/s]

Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw



/usr/local/lib/python3.7/dist-packages/torchvision/datasets/mnist.py:498: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


# Dataloader Arguments & Test/Train Dataloaders


In [ ]:
SEED = 1

# CUDA?
cuda = torch.cuda.is_available()
print("CUDA Available?", cuda)

# For reproducibility
torch.manual_seed(SEED)

if cuda:
    torch.cuda.manual_seed(SEED)

# dataloader arguments - something you'll fetch these from cmdprmt
dataloader_args = dict(shuffle=True, batch_size=128, num_workers=4, pin_memory=True) if cuda else dict(shuffle=True, batch_size=64)

# train dataloader
train_loader = torch.utils.data.DataLoader(train, **dataloader_args)

# test dataloader
test_loader = torch.utils.data.DataLoader(test, **dataloader_args)

CUDA Available? False


# The model
Let's start with the model we first saw

In [ ]:

dropout=.01
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        # Input Block
        self.convblock1 = nn.Sequential(
            nn.Conv2d(in_channels=1, out_channels=8, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(8),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 26


        # Input Block
        self.convblock2 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=8, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(8),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 24
        
        self.convblock3 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=8, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(8),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 22

        # TRANSITION BLOCK 1
        self.pool1 = nn.MaxPool2d(2, 2) # output_size = 11
        self.convblock4 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=13, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(13),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 9

        self.convblock5 = nn.Sequential(
            nn.Conv2d(in_channels=13, out_channels=14, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(14),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 7

        # OUTPUT BLOCK
        self.convblock6 = nn.Sequential(
            nn.Conv2d(in_channels=14, out_channels=8, kernel_size=(1, 1), padding=0, bias=False),
            nn.BatchNorm2d(8),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 7
         # OUTPUT BLOCK
        self.convblock7 = nn.Sequential(
            nn.Conv2d(in_channels=8, out_channels=26, kernel_size=(3, 3), padding=0, bias=False),
            nn.BatchNorm2d(26),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 5

        self.convblock8 = nn.Sequential(
            nn.Conv2d(in_channels=26, out_channels=50, kernel_size=(1, 1), padding=0, bias=False),
            nn.BatchNorm2d(50),
            nn.Dropout(dropout),
            nn.ReLU()
        ) # output_size = 5
        
        # OUTPUT BLOCK
        self.gap = nn.Sequential(
            nn.AvgPool2d(kernel_size=5)
        ) # out

        self.convblock9 = nn.Sequential(
            nn.Conv2d(in_channels=50, out_channels=10, kernel_size=(1, 1), padding=0, bias=False),
            # nn.BatchNorm2d(10),
            # nn.ReLU(),
            # nn.Dropout(dropout_value)
        ) 


    def forward(self, x):
        x = self.convblock1(x)
        x = self.convblock2(x)
        x = self.convblock3(x)
        x = self.pool1(x)
        x = self.convblock4(x)
        x = self.convblock5(x)
        x = self.convblock6(x)
        x = self.convblock7(x)
        x = self.convblock8(x)
        x = self.gap(x)
        x = self.convblock9(x)
        x = x.view(-1, 10)
        return F.log_softmax(x, dim=-1)

# Model Params
Can't emphasize on how important viewing Model Summary is. 
Unfortunately, there is no in-built model visualizer, so we have to take external help

In [ ]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print(device)
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

cpu
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
       BatchNorm2d-2            [-1, 8, 26, 26]              16
           Dropout-3            [-1, 8, 26, 26]               0
              ReLU-4            [-1, 8, 26, 26]               0
            Conv2d-5            [-1, 8, 24, 24]             576
       BatchNorm2d-6            [-1, 8, 24, 24]              16
           Dropout-7            [-1, 8, 24, 24]               0
              ReLU-8            [-1, 8, 24, 24]               0
            Conv2d-9            [-1, 8, 22, 22]             576
      BatchNorm2d-10            [-1, 8, 22, 22]              16
          Dropout-11            [-1, 8, 22, 22]               0
             ReLU-12            [-1, 8, 22, 22]               0
        MaxPool2d-13            [-1, 8, 11, 11]               0
           Conv2d-14             [-

# Training and Testing

All right, so we have 24M params, and that's too many, we know that. But the purpose of this notebook is to set things right for our future experiments. 

Looking at logs can be boring, so we'll introduce **tqdm** progressbar to get cooler logs. 

Let's write train and test functions

In [ ]:
from tqdm import tqdm

train_losses = []
test_losses = []
train_acc = []
test_acc = []

def train(model, device, train_loader, optimizer, epoch):
  model.train()
  pbar = tqdm(train_loader)
  correct = 0
  processed = 0
  for batch_idx, (data, target) in enumerate(pbar):
    # get samples
    data, target = data.to(device), target.to(device)

    # Init
    optimizer.zero_grad()
    # In PyTorch, we need to set the gradients to zero before starting to do backpropragation because PyTorch accumulates the gradients on subsequent backward passes. 
    # Because of this, when you start your training loop, ideally you should zero out the gradients so that you do the parameter update correctly.

    # Predict
    y_pred = model(data)

    # Calculate loss
    loss = F.nll_loss(y_pred, target)
    train_losses.append(loss)

    # Backpropagation
    loss.backward()
    optimizer.step()

    # Update pbar-tqdm
    
    pred = y_pred.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
    correct += pred.eq(target.view_as(pred)).sum().item()
    processed += len(data)

    pbar.set_description(desc= f'Loss={loss.item()} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')
    train_acc.append(100*correct/processed)

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    test_losses.append(test_loss)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))
    
    test_acc.append(100. * correct / len(test_loader.dataset))

In [ ]:

##dropout=.01  with scheduler scheduler = StepLR(optimizer, step_size=3, gamma=0.1)

from torch.optim.lr_scheduler import StepLR
model =  Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
scheduler = StepLR(optimizer, step_size=3, gamma=0.1)


EPOCHS = 15
for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    scheduler.step()
    test(model, device, test_loader)

EPOCH: 0


Loss=0.21597567200660706 Batch_id=937 Accuracy=94.07: 100%|██████████| 938/938 [01:08<00:00, 13.76it/s]



Test set: Average loss: 0.0835, Accuracy: 9748/10000 (97.48%)

EPOCH: 1


Loss=0.1222635880112648 Batch_id=937 Accuracy=97.97: 100%|██████████| 938/938 [01:07<00:00, 13.84it/s]



Test set: Average loss: 0.0354, Accuracy: 9875/10000 (98.75%)

EPOCH: 2


Loss=0.0020242019090801477 Batch_id=937 Accuracy=98.28: 100%|██████████| 938/938 [01:07<00:00, 13.92it/s]



Test set: Average loss: 0.0408, Accuracy: 9866/10000 (98.66%)

EPOCH: 3


Loss=0.08057547360658646 Batch_id=937 Accuracy=98.91: 100%|██████████| 938/938 [01:07<00:00, 13.82it/s]



Test set: Average loss: 0.0213, Accuracy: 9933/10000 (99.33%)

EPOCH: 4


Loss=0.06302107870578766 Batch_id=937 Accuracy=99.03: 100%|██████████| 938/938 [01:07<00:00, 13.83it/s]



Test set: Average loss: 0.0205, Accuracy: 9938/10000 (99.38%)

EPOCH: 5


Loss=0.028794988989830017 Batch_id=937 Accuracy=99.09: 100%|██████████| 938/938 [01:08<00:00, 13.72it/s]



Test set: Average loss: 0.0204, Accuracy: 9937/10000 (99.37%)

EPOCH: 6


Loss=0.001943838084116578 Batch_id=937 Accuracy=99.11: 100%|██████████| 938/938 [01:07<00:00, 13.81it/s]



Test set: Average loss: 0.0202, Accuracy: 9939/10000 (99.39%)

EPOCH: 7


Loss=0.17838214337825775 Batch_id=937 Accuracy=99.07: 100%|██████████| 938/938 [01:08<00:00, 13.73it/s]



Test set: Average loss: 0.0202, Accuracy: 9935/10000 (99.35%)

EPOCH: 8


Loss=0.00364412902854383 Batch_id=937 Accuracy=99.10: 100%|██████████| 938/938 [01:08<00:00, 13.74it/s]



Test set: Average loss: 0.0196, Accuracy: 9940/10000 (99.40%)

EPOCH: 9


Loss=0.00042675749864429235 Batch_id=937 Accuracy=99.06: 100%|██████████| 938/938 [01:08<00:00, 13.73it/s]



Test set: Average loss: 0.0200, Accuracy: 9935/10000 (99.35%)

EPOCH: 10


Loss=0.11231644451618195 Batch_id=937 Accuracy=99.11: 100%|██████████| 938/938 [01:08<00:00, 13.75it/s]



Test set: Average loss: 0.0210, Accuracy: 9937/10000 (99.37%)

EPOCH: 11


Loss=0.05786820873618126 Batch_id=937 Accuracy=99.09: 100%|██████████| 938/938 [01:07<00:00, 13.81it/s]



Test set: Average loss: 0.0197, Accuracy: 9939/10000 (99.39%)

EPOCH: 12


Loss=0.1500524878501892 Batch_id=937 Accuracy=99.12: 100%|██████████| 938/938 [01:08<00:00, 13.77it/s]



Test set: Average loss: 0.0200, Accuracy: 9938/10000 (99.38%)

EPOCH: 13


Loss=0.0026388473343104124 Batch_id=937 Accuracy=99.17: 100%|██████████| 938/938 [01:10<00:00, 13.39it/s]



Test set: Average loss: 0.0204, Accuracy: 9939/10000 (99.39%)

EPOCH: 14


Loss=0.04167485982179642 Batch_id=937 Accuracy=99.15: 100%|██████████| 938/938 [01:09<00:00, 13.52it/s]



Test set: Average loss: 0.0196, Accuracy: 9939/10000 (99.39%)



In [ ]:

##dropout=.01  with scheduler = StepLR(optimizer, step_size=2, gamma=0.5)


from torch.optim.lr_scheduler import StepLR
model =  Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
scheduler = StepLR(optimizer, step_size=2, gamma=0.5)


EPOCHS = 15
for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    scheduler.step()
    test(model, device, test_loader)

EPOCH: 0


Loss=0.018632998690009117 Batch_id=937 Accuracy=94.41: 100%|██████████| 938/938 [01:09<00:00, 13.56it/s]



Test set: Average loss: 0.0612, Accuracy: 9807/10000 (98.07%)

EPOCH: 1


Loss=0.05273209884762764 Batch_id=937 Accuracy=97.88: 100%|██████████| 938/938 [01:08<00:00, 13.61it/s]



Test set: Average loss: 0.0460, Accuracy: 9855/10000 (98.55%)

EPOCH: 2


Loss=0.08232365548610687 Batch_id=937 Accuracy=98.58: 100%|██████████| 938/938 [01:08<00:00, 13.61it/s]



Test set: Average loss: 0.0289, Accuracy: 9911/10000 (99.11%)

EPOCH: 3


Loss=0.006604798138141632 Batch_id=937 Accuracy=98.73: 100%|██████████| 938/938 [01:08<00:00, 13.65it/s]



Test set: Average loss: 0.0307, Accuracy: 9910/10000 (99.10%)

EPOCH: 4


Loss=0.010557949542999268 Batch_id=937 Accuracy=98.97: 100%|██████████| 938/938 [01:08<00:00, 13.62it/s]



Test set: Average loss: 0.0239, Accuracy: 9929/10000 (99.29%)

EPOCH: 5


Loss=0.009872930124402046 Batch_id=937 Accuracy=99.02: 100%|██████████| 938/938 [01:09<00:00, 13.56it/s]



Test set: Average loss: 0.0231, Accuracy: 9929/10000 (99.29%)

EPOCH: 6


Loss=0.19011856615543365 Batch_id=937 Accuracy=99.12: 100%|██████████| 938/938 [01:09<00:00, 13.56it/s]



Test set: Average loss: 0.0209, Accuracy: 9935/10000 (99.35%)

EPOCH: 7


Loss=0.08535661548376083 Batch_id=937 Accuracy=99.14: 100%|██████████| 938/938 [01:09<00:00, 13.44it/s]



Test set: Average loss: 0.0200, Accuracy: 9942/10000 (99.42%)

EPOCH: 8


Loss=0.007715530693531036 Batch_id=937 Accuracy=99.15: 100%|██████████| 938/938 [01:09<00:00, 13.52it/s]



Test set: Average loss: 0.0199, Accuracy: 9942/10000 (99.42%)

EPOCH: 9


Loss=0.04375467076897621 Batch_id=937 Accuracy=99.22: 100%|██████████| 938/938 [01:08<00:00, 13.62it/s]



Test set: Average loss: 0.0195, Accuracy: 9936/10000 (99.36%)

EPOCH: 10


Loss=0.006409833673387766 Batch_id=937 Accuracy=99.18: 100%|██████████| 938/938 [01:08<00:00, 13.65it/s]



Test set: Average loss: 0.0196, Accuracy: 9942/10000 (99.42%)

EPOCH: 11


Loss=0.0200789887458086 Batch_id=937 Accuracy=99.24: 100%|██████████| 938/938 [01:09<00:00, 13.52it/s]



Test set: Average loss: 0.0197, Accuracy: 9936/10000 (99.36%)

EPOCH: 12


Loss=0.0008217889117076993 Batch_id=937 Accuracy=99.29: 100%|██████████| 938/938 [01:09<00:00, 13.58it/s]



Test set: Average loss: 0.0192, Accuracy: 9939/10000 (99.39%)

EPOCH: 13


Loss=0.05975782871246338 Batch_id=937 Accuracy=99.28: 100%|██████████| 938/938 [01:08<00:00, 13.62it/s]



Test set: Average loss: 0.0192, Accuracy: 9943/10000 (99.43%)

EPOCH: 14


Loss=0.2638911008834839 Batch_id=937 Accuracy=99.28: 100%|██████████| 938/938 [01:09<00:00, 13.50it/s]



Test set: Average loss: 0.0188, Accuracy: 9944/10000 (99.44%)



In [ ]:

##dropout=.01  with scheduler = StepLR(optimizer, step_size=3, gamma=0.6)


from torch.optim.lr_scheduler import StepLR
model =  Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
scheduler = StepLR(optimizer, step_size=2, gamma=0.6)


EPOCHS = 15
for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    scheduler.step()
    test(model, device, test_loader)

EPOCH: 0


Loss=0.03843018040060997 Batch_id=937 Accuracy=94.07: 100%|██████████| 938/938 [01:09<00:00, 13.53it/s]



Test set: Average loss: 0.0908, Accuracy: 9721/10000 (97.21%)

EPOCH: 1


Loss=0.13390737771987915 Batch_id=937 Accuracy=97.77: 100%|██████████| 938/938 [01:09<00:00, 13.58it/s]



Test set: Average loss: 0.0649, Accuracy: 9796/10000 (97.96%)

EPOCH: 2


Loss=0.005649005062878132 Batch_id=937 Accuracy=98.55: 100%|██████████| 938/938 [01:09<00:00, 13.52it/s]



Test set: Average loss: 0.0290, Accuracy: 9912/10000 (99.12%)

EPOCH: 3


Loss=0.0650881826877594 Batch_id=937 Accuracy=98.73: 100%|██████████| 938/938 [01:08<00:00, 13.62it/s]



Test set: Average loss: 0.0342, Accuracy: 9891/10000 (98.91%)

EPOCH: 4


Loss=0.018602319061756134 Batch_id=937 Accuracy=98.92: 100%|██████████| 938/938 [01:08<00:00, 13.60it/s]



Test set: Average loss: 0.0208, Accuracy: 9937/10000 (99.37%)

EPOCH: 5


Loss=0.008030140772461891 Batch_id=937 Accuracy=98.91: 100%|██████████| 938/938 [01:08<00:00, 13.77it/s]



Test set: Average loss: 0.0245, Accuracy: 9921/10000 (99.21%)

EPOCH: 6


Loss=0.21844397485256195 Batch_id=937 Accuracy=99.09: 100%|██████████| 938/938 [01:08<00:00, 13.71it/s]



Test set: Average loss: 0.0194, Accuracy: 9940/10000 (99.40%)

EPOCH: 7


Loss=0.004417071118950844 Batch_id=937 Accuracy=99.02: 100%|██████████| 938/938 [01:08<00:00, 13.66it/s]



Test set: Average loss: 0.0185, Accuracy: 9943/10000 (99.43%)

EPOCH: 8


Loss=0.16748574376106262 Batch_id=937 Accuracy=99.19: 100%|██████████| 938/938 [01:08<00:00, 13.66it/s]



Test set: Average loss: 0.0201, Accuracy: 9935/10000 (99.35%)

EPOCH: 9


Loss=0.003505200147628784 Batch_id=937 Accuracy=99.23: 100%|██████████| 938/938 [01:08<00:00, 13.72it/s]



Test set: Average loss: 0.0189, Accuracy: 9935/10000 (99.35%)

EPOCH: 10


Loss=0.052215658128261566 Batch_id=937 Accuracy=99.25: 100%|██████████| 938/938 [01:10<00:00, 13.37it/s]



Test set: Average loss: 0.0194, Accuracy: 9935/10000 (99.35%)

EPOCH: 11


Loss=0.08852488547563553 Batch_id=937 Accuracy=99.27: 100%|██████████| 938/938 [01:10<00:00, 13.32it/s]



Test set: Average loss: 0.0186, Accuracy: 9946/10000 (99.46%)

EPOCH: 12


Loss=0.004165098536759615 Batch_id=937 Accuracy=99.30: 100%|██████████| 938/938 [01:10<00:00, 13.32it/s]



Test set: Average loss: 0.0178, Accuracy: 9945/10000 (99.45%)

EPOCH: 13


Loss=0.04324865713715553 Batch_id=937 Accuracy=99.27: 100%|██████████| 938/938 [01:10<00:00, 13.38it/s]



Test set: Average loss: 0.0174, Accuracy: 9942/10000 (99.42%)

EPOCH: 14


Loss=0.026008987799286842 Batch_id=937 Accuracy=99.32: 100%|██████████| 938/938 [01:10<00:00, 13.37it/s]



Test set: Average loss: 0.0177, Accuracy: 9943/10000 (99.43%)



1 - Target: Achive High Accuracy with less then 8K Parameters

2 - Results:

    1 - Parameters: 8k
    1 - Best Train Accuracy: 99.32 (At EPOCH - 14 , StepLR step_size=2, gamma=0.6)
    1 - Best Test Accuracy: 99.46  (At EPOCH - 11 , StepLR step_size=2, gamma=0.6)

3 - Analysis:

    We are able to achieve a high accuracy of 99.46 in the test dataset but
    this is slightly lesser than the high accuracy we achieve of 99.53 with
    param 10k . This is expected due to reduced model capacity .
  
